### PR TITLE
ref/Fail native ads when images fail to cache

### PR DESCRIPTION
**Pull request by santoshbagadi**

### CODE REVIEW

File: MoPub/iOS/AppLovinNativeCustomEvent.m [hunk 1]
```diff
@@ -69,16 +69,22 @@ - (void)nativeAdService:(ALNativeAdService *)service didLoadAds:(NSArray<ALNativ
...
+         else
+         {
+             [[self class] log: @"Native ad done precaching"];
+             AppLovinNativeAdapter *adapter = [[AppLovinNativeAdapter alloc] initWithNativeAd: nativeAd];
+             MPNativeAd *nativeAd = [[MPNativeAd alloc] initWithAdAdapter: adapter];
+             
              [self.delegate nativeCustomEvent: self didLoadAd: nativeAd];
-         });
-         
-         [adapter willAttachToView: nil];
```
Comment (thomasmso) [inline MoPub/iOS/AppLovinNativeCustomEvent.m]: Why'd we remove this callback?

File: MoPub/iOS/AppLovinNativeCustomEvent.m [hunk 2]
```diff
@@ -69,16 +69,22 @@ - (void)nativeAdService:(ALNativeAdService *)service didLoadAds:(NSArray<ALNativ
     
     [self precacheImagesWithURLs: imageURLs completionBlock:^(NSArray<NSError *> *errors)
      {
-         [[self class] log: @"Native ad done precaching"];
-         
-         AppLovinNativeAdapter *adapter = [[AppLovinNativeAdapter alloc] initWithNativeAd: nativeAd];
-         MPNativeAd *nativeAd = [[MPNativeAd alloc] initWithAdAdapter: adapter];
-         
-         dispatch_async(dispatch_get_main_queue(), ^{
+         if (errors.count > 0)
```
Comment (thomasmso) [inline MoPub/iOS/AppLovinNativeCustomEvent.m]: Could you keep to AppLovin coding style?

File: MoPub/iOS/AppLovinNativeCustomEvent.m [hunk 3]
```diff
@@ -69,16 +69,22 @@ - (void)nativeAdService:(ALNativeAdService *)service didLoadAds:(NSArray<ALNativ
     
     [self precacheImagesWithURLs: imageURLs completionBlock:^(NSArray<NSError *> *errors)
      {
-         [[self class] log: @"Native ad done precaching"];
-         
-         AppLovinNativeAdapter *adapter = [[AppLovinNativeAdapter alloc] initWithNativeAd: nativeAd];
-         MPNativeAd *nativeAd = [[MPNativeAd alloc] initWithAdAdapter: adapter];
-         
-         dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (thomasmso) [inline MoPub/iOS/AppLovinNativeCustomEvent.m]: Do you know if this is guaranteed to be ran on the main queue? MoPub doesn't like callbacks being called on anything else.

File: MoPub/iOS/AppLovinNativeCustomEvent.m (hunk 1 above)
Comment (santoshbagadi) [reply MoPub/iOS/AppLovinNativeCustomEvent.m]: The `willAttachToView:` will be called by MoPub when the ad view is shown by the publisher.

File: MoPub/iOS/AppLovinNativeCustomEvent.m (hunk 3 above)
Comment (santoshbagadi) [reply MoPub/iOS/AppLovinNativeCustomEvent.m]: It doesn't look like this is guaranteed. So adding back the change.

**Comment by santoshbagadi** [comment]
Changes made.

**Comment by santoshbagadi** [comment]
Added a few style changes.

**Comment by thomasmso** [comment]
LG2M, but lets run it through a formal round of QA during our next adapters push before merging.

File: MoPub/iOS/AppLovinNativeCustomEvent.m [hunk 4]
```diff
@@ -69,16 +69,27 @@ - (void)nativeAdService:(ALNativeAdService *)service didLoadAds:(NSArray<ALNativ
...
-         
-         [adapter willAttachToView: nil];
+         if ( errors.count > 0 )
+         {
+             [[self class] log: @"Native ad failed to precache images."];
+             NSError *error = [NSError errorWithDomain: kALMoPubMediationErrorDomain
+                                                  code: MPNativeAdErrorImageDownloadFailed
+                                              userInfo: nil];
+             
+             dispatch_sync(dispatch_get_main_queue(), ^{
```
Comment (christophercong) [inline MoPub/iOS/AppLovinNativeCustomEvent.m]: This can lock. Let's use `dispatch_async`

File: MoPub/iOS/AppLovinNativeCustomEvent.m [hunk 5]
```diff
@@ -69,16 +69,27 @@ - (void)nativeAdService:(ALNativeAdService *)service didLoadAds:(NSArray<ALNativ
...
+                 [self.delegate nativeCustomEvent: self didFailToLoadAdWithError: error];
+             })
+         }
+         else
+         {
+             [[self class] log: @"Native ad done precaching"];
+             AppLovinNativeAdapter *adapter = [[AppLovinNativeAdapter alloc] initWithNativeAd: nativeAd];
+             MPNativeAd *nativeAd = [[MPNativeAd alloc] initWithAdAdapter: adapter];
+             
+             dispatch_sync(dispatch_get_main_queue(), ^{
```
Comment (christophercong) [inline MoPub/iOS/AppLovinNativeCustomEvent.m]: Same comment as above.

**Comment by santoshbagadi** [comment]
Changes made.

File: MoPub/Android/AppLovinCustomEventNative.java [hunk 6]
```diff
@@ -179,6 +160,17 @@ public void onPostbackFailure(String url, int errorCode)
...
+            nativeClickHandler.setOnClickListener( view, this );
+        }
+
+        @Override
+        public void handleClick(@NonNull final View view)
+        {
+            super.handleClick( view );
+
+            parentNativeAd.launchClickTarget( view.getContext() );
+            notifyAdClicked();
```
Comment (thomasmso) [inline MoPub/Android/AppLovinCustomEventNative.java]: Can we keep the order we have previously? That is 1. notify ad click, 2. launch click target.

Comment (santoshbagadi) [reply MoPub/Android/AppLovinCustomEventNative.java]: This was the order that was being used previously. But I think it makes sense to update the order.

**Comment by santoshbagadi** [comment]
Changes made.

**Comment by thomasmso** [comment]
@santoshbagadi - I've changed the base branch to `release/admob_2_3_3_and_mopub_3_0_1` - feel free to merge it into that.

**Pull request merged**